src/Serializer/AbstractItemNormalizer.php

```diff
@@ -183,14 +183,33 @@ protected function setAttributeValue($object, $attribute, $value, $format = null
             return;
         }
 
+        $this->validateType($attribute, $type, $value, $format);
+        $this->setValue($object, $attribute, $value);
+    }
+
+    /**
+     * Validates the type of the value. Allows using integers as floats for JSON formats.
+     *
+     * @param string      $attribute
+     * @param Type        $type
+     * @param string|null $format
+     *
+     * @throws InvalidArgumentException
+     */
+    protected function validateType(string $attribute, Type $type, $value, string $format = null)
+    {
         $builtinType = $type->getBuiltinType();
-        if (!call_user_func('is_'.$builtinType, $value)) {
+        if (false !== strpos($format, 'json') && Type::BUILTIN_TYPE_FLOAT === $builtinType) {
```

Comment:
No, it's a specificity of JSON to have a generic `Number` type. Other formats generally make the distinction between `int` and `float`.